server/task/src/main/java/org/apache/james/task/MemoryTaskManager.java

```diff
@@ -20,101 +20,57 @@
 package org.apache.james.task;
 
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.concurrent.Future;
-import java.util.concurrent.ThreadFactory;
 import java.util.function.Consumer;
+import java.util.function.Function;
 
 import javax.annotation.PreDestroy;
 
-import org.apache.james.util.MDCBuilder;
-import org.apache.james.util.concurrent.NamedThreadFactory;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 import com.github.fge.lambdas.Throwing;
 import com.github.steveash.guavate.Guavate;
-import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.WorkQueueProcessor;
+import reactor.core.scheduler.Schedulers;
 
 public class MemoryTaskManager implements TaskManager {
-    private static final boolean INTERRUPT_IF_RUNNING = true;
-    private static final Logger LOGGER = LoggerFactory.getLogger(MemoryTaskManager.class);
-
+    private final WorkQueueProcessor<TaskWithId> workQueue;
+    private final TaskManagerWorker worker;
     private final ConcurrentHashMap<TaskId, TaskExecutionDetails> idToExecutionDetails;
-    private final ConcurrentHashMap<TaskId, Future<?>> idToFuture;
-    private final ExecutorService executor;
+    private final ConcurrentHashMap<TaskId, Mono<Task.Result>> tasksResult;
 
     public MemoryTaskManager() {
+        workQueue = WorkQueueProcessor.<TaskWithId>builder()
+            .executor(Executors.newSingleThreadExecutor())
+            .requestTaskExecutor(Executors.newSingleThreadExecutor())
+            .build();
         idToExecutionDetails = new ConcurrentHashMap<>();
-        idToFuture = new ConcurrentHashMap<>();
-        ThreadFactory threadFactory = NamedThreadFactory.withClassName(getClass());
-        executor = Executors.newSingleThreadExecutor(threadFactory);
+        tasksResult = new ConcurrentHashMap<>();
+        worker = new MemoryTaskManagerWorker();
+        workQueue
+            .subscribeOn(Schedulers.single())
+            .filter(task -> !listIds(Status.CANCELLED).contains(task.getId()))
+            .subscribe(this::treatTask);
     }
 
-    @Override
-    public TaskId submit(Task task) {
-        return submit(task, id -> { });
+    private void treatTask(TaskWithId task) {
+        Mono<Task.Result> result = worker.executeTask(task, updateDetails(task.getId())).cache();
```

Comment:
why do you cache()?

Reply:
It was in order to memoize the result but it's not really needed here.